src/Dictionaries/HashedDictionaryCollectionType.h

```diff
@@ -0,0 +1,218 @@
+#pragma once
+
+#include <Dictionaries/IDictionary.h>
+#include <Common/HashTable/PackedHashMap.h>
+#include <Common/HashTable/HashMap.h>
+#include <Common/HashTable/HashSet.h>
+#include <sparsehash/sparse_hash_map>
+#include <sparsehash/sparse_hash_set>
+#include <type_traits>
+
+namespace DB
+{
+
+/// HashMap with packed structure is better then google::sparse_hash_map if the
+/// <K, V> pair is small, for the sizeof(std::pair<K, V>) == 16, RSS for hash
+/// table with 1e9 elements will be:
+///
+/// - google::sparse_hash_map             : 26GiB
+/// - HashMap                             : 35GiB
+/// - PackedHashMap                       : 22GiB
+/// - google::sparse_hash_map<packed_pair>: 17GiB
+///
+/// Also note here sizeof(std::pair<>) was used since google::sparse_hash_map
+/// uses it to store <K, V>, yes we can modify google::sparse_hash_map to work
+/// with packed analog of std::pair, but the allocator overhead is still
+/// significant, because of tons of reallocations (and those cannot be solved
+/// with reserve() due to some internals of google::sparse_hash_map) and poor
+/// jemalloc support of such pattern, which results in 33% fragmentation (in
+/// comparison with glibc).
+///
+/// Plus since google::sparse_hash_map cannot use packed structure, it will
+/// have the same memory footprint for everything from UInt8 to UInt64 values
+/// and so on.
+///
+/// Returns true hen google::sparse_hash_map should be used, otherwise
+/// PackedHashMap should be used instead.
+template <typename K, typename V>
+constexpr bool useSparseHashForHashedDictionary()
+{
+    return sizeof(PackedPairNoInit<K, V>) > 16;
+}
+
+/// Grower with custom fill limit/load factor (instead of default 50%).
+///
+/// It turns out that HashMap can outperform google::sparse_hash_map in case of
+/// the structure size of not big, in terms of speed *and* memory. Even 99% of
+/// max load factor was faster then google::sparse_hash_map in my simple tests
+/// (1e9 UInt64 keys with UInt16 values, randomly distributed).
+///
+/// And not to mention very high allocator memory fragmentation in
+/// google::sparse_hash_map.
+///
+/// Based on HashTableGrowerWithPrecalculation
+class alignas(64) HashTableGrowerWithMaxLoadFactor
+{
+    static constexpr size_t initial_size_degree = 8;
+    UInt8 size_degree = initial_size_degree;
+    size_t precalculated_mask = (1ULL << initial_size_degree) - 1;
+    size_t precalculated_max_fill = 1ULL << (initial_size_degree - 1);
+    float max_load_factor = 0.5;
+    /// HashTableGrowerWithPrecalculation has 23
+    size_t max_size_degree_quadratic = 19;
```

Comment:
>a) if 19 was found experimentally (if so, let's add a comment) <- probably yes, as "23" in HTGrowerWithPrecalculation looks equally "random"

This is just to decrease memory usage slightly and I doubt that there is huge difference (in terms of speed) between 23 and 19, especially for dictionaries (since they are not that frequently loaded).

>for symmetry, the corresponding constant in HashTableGrowerWithPrecalculation could be pulled out as max_size_degree.

OK, will do (though I think that touching extra bits in the patches makes review process harder.)